File: Remote/Source/Remote Elements/Views/ButtonView.m
//
// ButtonView.m
// Remote
//
// Created by Jason Cardwell on 5/24/11.
// Copyright 2011 Moondeer Studios. All rights reserved.
//
#import "RemoteElementView_Private.h"
#import "ImageView.h"
#import "TitleAttributes.h"
#import "Button.h"
#import "Command.h"
#import "ControlStateTitleSet.h"

// #define DEBUG_BV_COLOR_BG

static int       ddLogLevel   = LOG_LEVEL_DEBUG;
static const int msLogContext = (LOG_CONTEXT_REMOTE | LOG_CONTEXT_FILE | LOG_CONTEXT_CONSOLE);
#pragma unused(ddLogLevel, msLogContext)


@interface ButtonView ()
@property (nonatomic, weak,   readwrite)  UITapGestureRecognizer       * tapGesture;
@property (nonatomic, weak,   readwrite)  MSLongPressGestureRecognizer * longPressGesture;
@property (nonatomic, weak,   readwrite)  UILabel                      * labelView;
@property (nonatomic, weak,   readwrite)  UIActivityIndicatorView      * activityIndicator;
@end

@implementation ButtonView


#pragma mark Internal subviews and constraints
////////////////////////////////////////////////////////////////////////////////


- (void)addInternalSubviews {
  [super addInternalSubviews];

  self.subelementInteractionEnabled = NO;
  self.contentInteractionEnabled    = NO;

  UILabel * labelView = [UILabel newForAutolayout];
  [self addViewToContent:labelView];
  self.labelView = labelView;

  UIActivityIndicatorView * activityIndicator = [UIActivityIndicatorView newForAutolayout];
  activityIndicator.activityIndicatorViewStyle = UIActivityIndicatorViewStyleWhiteLarge;
  activityIndicator.color                      = defaultTitleHighlightColor();
  [self addViewToOverlay:activityIndicator];
  self.activityIndicator = activityIndicator;
}

- (void)updateConstraints {
  [super updateConstraints];

  NSString * labelNametag    = ClassNametagWithSuffix(@"InternalLabel");
  NSString * activityNametag = ClassNametagWithSuffix(@"InternalActivity");

  if (![self constraintsWithNametagPrefix:labelNametag]) {
    UIEdgeInsets titleInsets = self.model.titleEdgeInsets;
    NSString   * constraints =
      $(@"'%1$@' label.left = self.left + %3$f @900\n"
        "'%1$@' label.top = self.top + %4$f @900\n"
        "'%1$@' label.bottom = self.bottom - %5$f @900\n"
        "'%1$@' label.right = self.right - %6$f @900\n"
        "'%2$@' activity.centerX = self.centerX\n"
        "'%2$@' activity.centerY = self.centerY",
        labelNametag, activityNametag,
        titleInsets.left, titleInsets.top, titleInsets.bottom, titleInsets.right);

    NSDictionary * views = @{@"self": self, @"label": self.labelView, @"activity": self.activityIndicator};

    [self addConstraints:[NSLayoutConstraint constraintsByParsingString:constraints views:views]];
  }
}

#pragma mark Gestures
////////////////////////////////////////////////////////////////////////////////


- (void)attachGestureRecognizers {
  [super attachGestureRecognizers];

  MSLongPressGestureRecognizer * longPressGesture =
    [MSLongPressGestureRecognizer gestureWithTarget:self action:@selector(handleLongPress:)];

  longPressGesture.delaysTouchesBegan = NO;
  longPressGesture.delegate           = self;
  [self addGestureRecognizer:longPressGesture];
  self.longPressGesture = longPressGesture;

  UITapGestureRecognizer * tapGesture =
    [UITapGestureRecognizer gestureWithTarget:self action:@selector(handleTap:)];

  tapGesture.numberOfTapsRequired    = 1;
  tapGesture.numberOfTouchesRequired = 1;
  tapGesture.delaysTouchesBegan      = NO;
  tapGesture.delegate                = self;
  [self addGestureRecognizer:tapGesture];
  self.tapGesture = tapGesture;
}

/// Single tap action executes the primary button command
- (void)handleTap:(UITapGestureRecognizer *)gestureRecognizer {

  if (gestureRecognizer.state == UIGestureRecognizerStateRecognized) {

    self.model.highlighted = YES;

    __weak ButtonView * weakself = self;
    MSDelayedRunOnMain(1.0, ^{ weakself.model.highlighted = NO; });

    if (self.tapAction) self.tapAction();
    else [self executeActionWithOptions:CommandOptionDefault];
  }
}

/// Long press action executes the secondary button command
- (void)handleLongPress:(MSLongPressGestureRecognizer *)gestureRecognizer {

  if (gestureRecognizer.state == UIGestureRecognizerStateBegan) {

    if (self.pressAction) self.pressAction();
    else [self executeActionWithOptions:CommandOptionLongPress];

  } else if (gestureRecognizer.state == UIGestureRecognizerStatePossible) {
    self.model.highlighted = YES;
    [self setNeedsDisplay];
  }
}


#pragma mark Button actions
////////////////////////////////////////////////////////////////////////////////


- (void)executeActionWithOptions:(CommandOptions)options {

  if (!self.editing) {

    if (self.model.command.indicator) [_activityIndicator startAnimating];

    [self.model executeCommandWithOptions:options
                               completion:^(BOOL success, NSError * error) {
                                 if ([self.activityIndicator isAnimating])
                                   MSRunAsyncOnMain(^{ [self.activityIndicator stopAnimating]; });
                               }];
  }

}

#pragma mark Content size
////////////////////////////////////////////////////////////////////////////////


- (CGSize)intrinsicContentSize { return self.minimumSize; }

- (CGSize)minimumSize {

  CGRect frame = (CGRect) { .size = REMinimumSize };

  NSMutableSet * titles = [NSMutableSet set];

  for (NSString *mode in self.model.modes) {
    ControlStateTitleSet * titleSet = [self.model titlesForMode:mode];
    if (titleSet) [titles addObjectsFromArray:[titleSet allValues]];
  }

  if ([titles count]) {

    CGFloat maxWidth = 0.0, maxHeight = 0.0;

    for (NSAttributedString * title in [titles valueForKeyPath:@"string"]) {

      CGSize titleSize = [title size];
      UIEdgeInsets titleInsets = self.model.titleEdgeInsets;

      titleSize.width  += titleInsets.left + titleInsets.right;
      titleSize.height += titleInsets.top + titleInsets.bottom;

      maxWidth = MAX(titleSize.width, maxWidth);
      maxHeight = MAX(titleSize.height, maxHeight);

    }

    frame.size.width = MAX(maxWidth, frame.size.width);
    frame.size.height = MAX(maxHeight, frame.size.height);
  }

  if (self.model.proportionLock && !CGSizeEqualToSize(self.bounds.size, CGSizeZero)) {

    CGSize currentSize = self.bounds.size;

    if (currentSize.width > currentSize.height)
      frame.size.height = (frame.size.width * currentSize.height) / currentSize.width;

    else
      frame.size.width = (frame.size.height * currentSize.width) / currentSize.height;
  }

  return frame.size;
}

#pragma mark Prevent adding subelement views
////////////////////////////////////////////////////////////////////////////////


- (void)addSubelementView:(RemoteElementView *)view {}
- (void)removeSubelementView:(RemoteElementView *)view {}
- (void)addSubelementViews:(NSSet *)views {}
- (void)removeSubelementViews:(NSSet *)views {}
- (NSArray *)subelementViews { return nil; }


#pragma mark Initialization
////////////////////////////////////////////////////////////////////////////////


- (void)initializeIVARs {
  self.cornerRadii    = (CGSize) { 5.0f, 5.0f };
  [super initializeIVARs];
}

- (void)initializeViewFromModel {
  [super initializeViewFromModel];

  self.longPressGesture.enabled = (self.model.longPressCommand != nil);
  self.labelView.attributedText = self.model.title;

  [self invalidateIntrinsicContentSize];
  [self setNeedsDisplay];
}

#pragma mark Key-value observing
////////////////////////////////////////////////////////////////////////////////


- (MSDictionary *)kvoRegistration {

  MSDictionary * reg = [super kvoRegistration];

//    reg[@"selected"] = ^(MSKVOReceptionist * receptionist) {
//      [(__bridge ButtonView *)receptionist.context updateState];
//    };

//    reg[@"enabled"] = ^(MSKVOReceptionist * receptionist) {
//      ButtonView * buttonView = (__bridge ButtonView *)receptionist.context;
//      buttonView.enabled = [receptionist.change[NSKeyValueChangeNewKey] boolValue];
//    };

//    reg[@"highlighted"] = ^(MSKVOReceptionist * receptionist) {
//      [(__bridge ButtonView *)receptionist.context setNeedsDisplay];
//    };

    reg[@"title"] = ^(MSKVOReceptionist * receptionist) {
      ButtonView         * buttonView = (ButtonView *)receptionist.observer;
      NSAttributedString * title      = NilSafe(receptionist.change[NSKeyValueChangeNewKey]);
      buttonView.labelView.attributedText = title;
      [buttonView invalidateIntrinsicContentSize];
      [buttonView setNeedsDisplay];
    };

//    reg[@"image"] = ^(MSKVOReceptionist * receptionist) {
//      assert(NO);
//    };

    reg[@"icon"] = ^(MSKVOReceptionist * receptionist) {
      ButtonView * buttonView = (ButtonView *)receptionist.observer;
      [buttonView invalidateIntrinsicContentSize];
      [buttonView setNeedsDisplay];
    };

  return reg;
}

#pragma mark Editing
////////////////////////////////////////////////////////////////////////////////


- (void)setEditingMode:(REEditingMode)editingMode {
  [super setEditingMode:editingMode];
  BOOL enabled = !self.isEditing;
  self.tapGesture.enabled       = enabled;
  self.longPressGesture.enabled = enabled;
}

#pragma mark Drawing
////////////////////////////////////////////////////////////////////////////////


- (void)drawContentInContext:(CGContextRef)ctx inRect:(CGRect)rect {
  UIImage * icon = self.model.icon.colorImage;
  if (icon) {
    UIGraphicsPushContext(ctx);
    CGRect insetRect = UIEdgeInsetsInsetRect(self.bounds, self.model.imageEdgeInsets);
    CGSize imageSize = (CGSizeContainsSize(insetRect.size, icon.size)
                        ? icon.size
                        : CGSizeAspectMappedToSize(icon.size, insetRect.size, YES));
    CGRect imageRect = CGRectMake(CGRectGetMidX(insetRect) - imageSize.width / 2.0,
                                  CGRectGetMidY(insetRect) - imageSize.height / 2.0,
                                  imageSize.width,
                                  imageSize.height);

    [icon drawInRect:imageRect];
    UIGraphicsPopContext();
  }
}

@end

File: Remote/Source/Remote Elements/Views/ButtonView.swift
//
//  ButtonView.swift
//  Remote
//
//  Created by Jason Cardwell on 11/07/14.
//  Copyright (c) 2014 Moondeer Studios. All rights reserved.
//

import Foundation
import UIKit
import MoonKit

class ButtonView: RemoteElementView {

	weak var tapGesture :UITapGestureRecognizer!
	weak var longPressGesture: MSLongPressGestureRecognizer!
	weak var labelView: UILabel!
	weak var activityIndicator: UIActivityIndicatorView!

  var tapAction: ((Void) -> Void)?
  var pressAction: ((Void) -> Void)?
  var button: Button { return model as Button }

  /**
  viewWithModel:

  :param: model Button

  :returns: ButtonView
  */
  @objc(viewWithButton:)
  override class func viewWithModel(model: Button) -> ButtonView {
    switch model.role {
      case RemoteElement.Role.BatteryStatus:    return BatteryStatusButtonView(model: model)
      case RemoteElement.Role.ConnectionStatus: return ConnectionStatusButtonView(model: model)
      default:                                return ButtonView(model: model)
    }
  }

  /** init */
  override init() { super.init() }

  /**
  initWithFrame:

  :param: frame CGRect
  */
  override init(frame: CGRect) { super.init(frame: frame) }

  /**
  Overridden properties prevent synthesized initializers

  :param: model RemoteElement
  */
  required init(model: RemoteElement) { super.init(model: model) }

  /**
  Overridden properties prevent synthesized initializers

  :param: aDecoder NSCoder
  */
  required init(coder aDecoder: NSCoder) { fatalError("init(coder:) has not been implemented") }

	/**
	addSubelementView:

	:param: view RemoteElementView
	*/
	override func addSubelementView(view: RemoteElementView) {}

	/**
	removeSubelementView:

	:param: view RemoteElementView
	*/
	override func removeSubelementView(view: RemoteElementView) {}

	/**
	addSubelementViews:

	:param: views NSSet
	*/
	override func addSubelementViews(views: NSSet) {}

	/**
	removeSubelementViews:

	:param: views NSSet
	*/
	override func removeSubelementViews(views: NSSet) {}

	override var subelementViews: [RemoteElementView] { return [] }

	/**
	executeActionWithOptions:

	:param: options CommandOptions
	*/
	func executeActionWithOptions(options: CommandOptions) {
		if !isEditing {
			if button.command != nil && button.command!.indicator { activityIndicator.startAnimating() }
			button.executeCommandWithOptions(options) {
				(success: Bool, error: NSError?) -> Void in
					if self.activityIndicator.isAnimating() {
						NSOperationQueue.mainQueue().addOperationWithBlock {
							self.activityIndicator.stopAnimating()
						}
					}
			}
		}
	}

	/** attachGestureRecognizers */
	override func attachGestureRecognizers() {
		super.attachGestureRecognizers()

		let longPressGesture = MSLongPressGestureRecognizer(target: self, action: "handleLongPress:")
		longPressGesture.delaysTouchesBegan = false
		addGestureRecognizer(longPressGesture)
		self.longPressGesture = longPressGesture

		let tapGesture = UITapGestureRecognizer(target: self, action: "handleTap:")
		tapGesture.numberOfTapsRequired    = 1
		tapGesture.numberOfTouchesRequired = 1
		tapGesture.delaysTouchesBegan      = false
		addGestureRecognizer(tapGesture)
		self.tapGesture = tapGesture
	}

	/**
	Single tap action executes the primary button command

	:param: gestureRecognizer UITapGestureRecognizer
	*/
	func handleTap(gestureRecognizer: UITapGestureRecognizer) {
		if gestureRecognizer.state == .Ended {
			button.highlighted = true
			MSDelayedRunOnMain(1, {self.button.highlighted = false})
			tapAction?() ?? executeActionWithOptions(.Default)
		}
	}

	/**
	Long press action executes the secondary button command

	:param: gestureRecognizer MSLongPressGestureRecognizer
	*/
	func handleLongPress(gestureRecognizer: MSLongPressGestureRecognizer) {
		if gestureRecognizer.state == .Ended {
			pressAction?() ?? executeActionWithOptions(.LongPress)

		} else if gestureRecognizer.state == .Possible {
			button.highlighted = true
			setNeedsDisplay()
		}
	}

	/** addInternalSubviews */
	override func addInternalSubviews() {
		super.addInternalSubviews()

		subelementInteractionEnabled = false
		contentInteractionEnabled    = false

		let labelView = UILabel.newForAutolayout()
		addViewToContent(labelView)
		self.labelView = labelView

		let activityIndicator = UIActivityIndicatorView.newForAutolayout()
		activityIndicator.activityIndicatorViewStyle = .WhiteLarge
		activityIndicator.color = UIColor.whiteColor()
		addViewToOverlay(activityIndicator)
		self.activityIndicator = activityIndicator
	}

	/** updateConstraints */
	override func updateConstraints() {
		super.updateConstraints()

    let identifier = createIdentifier(self, ["Button", "Internal"])
    if constraintsWithIdentifier(identifier).count == 0 {
      let titleInsets = button.titleEdgeInsets
      let format = "\n".join("label.left = self.left + \(titleInsets.left) @900",
                             "label.top = self.top + \(titleInsets.top) @900",
                             "label.bottom = self.bottom - \(titleInsets.bottom) @900",
                             "label.right = self.right - \(titleInsets.right) @900",
                             "activity.center = self.center")
      constrain(format, views: ["label": labelView, "activity": activityIndicator], identifier: identifier)
    }
	}


 /**
	intrinsicContentSize

	:returns: CGSize
	*/
	override func intrinsicContentSize() -> CGSize { return minimumSize }

	override var minimumSize: CGSize {
		var frame = CGRect(size: RemoteElementView.MinimumSize)
//	  let titleSets = model.modes.map{self.button.titlesForMode($0 as NSString)}
//	  let titles = titleSets.map{$0.allValues}
//
//	  if titles.count > 0 {
//	  	var maxWidth: CGFloat = 0.0
//	  	var maxHeight: CGFloat = 0.0
//
//	  	for title in (titles.map{$0.string}) {
//	  		var titleSize = title.size
//	  		let titleInsets = model.titleEdgeInsets
//	  		titleSize.width += titleInsets.left + titleInsets.right
//	  		titleSize.height += titleInsets.top + titleInsets.bottom
//	  		maxWidth = max(titleSize.width, maxWidth)
//	  		maxHeight = max(titleSize.height, maxHeight)
//	  	}
//
//	    frame.size.width = max(maxWidth, frame.width)
//	    frame.size.height = max(maxHeight, frame.height)
//	  }
//
    if let title = button.title {
      var titleSize = title.size()
      let titleInsets = button.titleEdgeInsets
      titleSize.width += titleInsets.left + titleInsets.right
      titleSize.height += titleInsets.top + titleInsets.bottom
      frame.size.width = max(titleSize.width, frame.size.width)
      frame.size.height = max(titleSize.height, frame.size.height)
    }

	  if model.constraintManager.proportionLock && bounds.size != CGSize.zeroSize {
	  	let currentSize = bounds.size

	    if currentSize.width > currentSize.height {
	    	frame.size.height = (frame.size.width * currentSize.height) / currentSize.width
    	} else {
    		frame.size.width = (frame.size.height * currentSize.width) / currentSize.height
    	}
    }

	  return frame.size
	}

	/**
	kvoRegistration

	:returns: [String:(MSKVOReceptionist) -> Void]
	*/
	override func kvoRegistration() -> [String:(MSKVOReceptionist!) -> Void] {
		var registry = super.kvoRegistration()
		registry["title"] = {
			(receptionist: MSKVOReceptionist!) -> Void in
				if let v = receptionist.observer as? ButtonView {
					v.labelView.attributedText = receptionist.change[NSKeyValueChangeNewKey] as? NSAttributedString
					v.invalidateIntrinsicContentSize()
					v.setNeedsDisplay()
				}
		}
		registry["icon"] = {
			(receptionist: MSKVOReceptionist!) -> Void in
				if let v = receptionist.observer as? ButtonView {
					v.invalidateIntrinsicContentSize()
					v.setNeedsDisplay()
				}
		}
		return registry
	}

	/** initializeViewFromModel */
	override func initializeViewFromModel() {
		super.initializeViewFromModel()
		longPressGesture.enabled = button.longPressCommand != nil
		labelView.attributedText = button.title
		invalidateIntrinsicContentSize()
		setNeedsDisplay()
	}

	override var editingMode: RemoteElement.BaseType {
		didSet {
			tapGesture.enabled = !isEditing
			longPressGesture.enabled = !isEditing
		}
	}

	/**
	drawContentInContext:inRect:

	:param: ctx CGContextRef
	:param: rect CGRect
	*/
	override func drawContentInContext(ctx: CGContextRef, inRect rect: CGRect) {
		if let icon = button.icon?.colorImage {
			UIGraphicsPushContext(ctx)
			let insetRect = button.imageEdgeInsets.insetRect(bounds)
      let imageSize = insetRect.size.contains(icon.size)
                        ? icon.size
                        : icon.size.aspectMappedToSize(insetRect.size, binding: true)
			let imageRect = CGRect(x: insetRect.midX - imageSize.width / 2.0,
				                     y: insetRect.midY - imageSize.height / 2.0,
				                     width: imageSize.width,
				                     height: imageSize.height)
			icon.drawInRect(imageRect)
			UIGraphicsPopContext()
		}

	}

}
